### PR TITLE
fix(deepseek_v4): support packed THD with context parallel

### LINKED ISSUE #2878: DSV4 THD packed + CP
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by HuiyingLi**
## What

Support DeepSeek V4 packed THD training with model-owned context parallelism.

This PR:

- wires DeepSeek V4 into the CP-capable model path
- adds DSV4 CP helpers that shard each rank over a contiguous query slice while all-gathering K/V and compressed K/V for attention
- pads packed documents to the DSV4 compression-safe local multiple before CP slicing
- preserves packed metadata needed by attention, including `seq_lens`, `seq_lens_padded`, `position_ids`, `padding_mask`, and `packed_seq_ids`
- updates DSV4 sparse/compressed attention indexing to use global query/key positions under CP

## Why

The generic packed-sequence path can hand DSV4 a THD batch whose document boundaries do not line up with the compression windows used by CSA/HCA. Once CP slices the sequence, those boundaries also affect global query positions, compressed K/V visibility, and padding masks.

DSV4 needs to own the packed CP batch preparation so document-level padding is inserted before the CP split and the attention code can rebuild local-query/global-key masks after K/V all-gather.

## Validation

- `python -m pytest -q tests/unit_tests/models/deepseek_v4/test_dsv4_cp_batch.py tests/unit_tests/models/deepseek_v4/test_dsv4_layers.py -k 'cp or packed_topk or eager_attention_with_sink_matches_unpacked_packed_sequence'`
  - `44 passed, 2 skipped, 51 deselected`
- `ruff check nemo_automodel/_transformers/capabilities.py nemo_automodel/components/distributed/cp_utils.py nemo_automodel/components/models/deepseek_v4/cp.py nemo_automodel/components/models/deepseek_v4/fsdp.py nemo_automodel/components/models/deepseek_v4/layers.py nemo_automodel/components/models/deepseek_v4/model.py tests/unit_tests/models/deepseek_v4/test_dsv4_cp_batch.py`
  - `All checks passed`
- Large DSV4 packed THD + CP validation completed 100/100 optimizer steps with finite loss and grad norm:
  - https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/2suf2cl1?nw=nwuserhuiyingl
- GLM5.2 packed THD + CP1 validation with HybridEP (PP4/EP64) completed successfully:
  - https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/pssq9x60

image

### CODE REVIEW

**Comment by akoumpa** [comment]
@HuiyingLi is this working with the latest 26.06 rc?

**Comment by HuiyingLi** [comment]
@akoumpa  This recipe is tested with 2606rc9 container with hybridep

File: nemo_automodel/_transformers/capabilities.py [hunk 1]
```diff
@@ -273,16 +273,16 @@ def supports_ep(self) -> bool:
...
-            or _uses_magi_attention(self._model)
-            or (
-                _is_glm_moe_dsa(self._model)
-                and getattr(getattr(self._model, "backend", None), "attn", None) == "tilelang"
-            )
+            getattr(model, "_supports_sdpa", False) is True
+            or _uses_te_attention(model)
+            or _uses_magi_attention(model)
+            or (_is_deepseek_v4(model) and backend_attn == "tilelang")
+            or (_is_glm_moe_dsa(model) and backend_attn == "tilelang")
```
Comment (akoumpa) [inline nemo_automodel/_transformers/capabilities.py]: @HuiyingLi instead of this, can you add a `supports_thd` field in the capabilities struct instead and fetch it from there?

**Comment by akoumpa** [comment]
Thank you @HuiyingLi for this great feature; do you have non-thd vs thd loss graph to compare? thank you

**Comment by HuiyingLi** [comment]
@akoumpa https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/n0j0v1ga

**Comment by HuiyingLi** [comment]
@akoumpa 
image

File: nemo_automodel/recipes/llm/train_ft.py [hunk 2]
```diff
@@ -1291,16 +1296,6 @@ def _forward_backward_step(
...
-                # Model-owned CP (e.g. DSV4) can request a zero-valued full-logits
-                # term so every CP rank's backward reaches all parameters even when
-                # its local loss is fully masked (avoids FSDP2 unused-parameter hangs).
-                if is_train and batch.get("_cp_full_logits_grad_touch"):
-                    logits = getattr(out, "logits", out)
-                    if isinstance(logits, torch.Tensor):
-                        # Promote to fp32 before summing: bf16 logits over a large
-                        # vocab (e.g. DSV4's 129280) overflow to inf, and inf * 0.0
-                        # would be nan, poisoning local_loss and the backward pass.
-                        local_loss = local_loss + logits.float().sum() * 0.0
```
Comment (hemildesai) [inline nemo_automodel/recipes/llm/train_ft.py:1303]: why is this removed? does it affect other models?

Comment (HuiyingLi) [reply nemo_automodel/recipes/llm/train_ft.py:1303]: Good question. This was originally introduced during the DSV4 CP work as a defensive adaptation of Miles’ `0 * logits.sum()` workaround. Miles had observed a GLM-5 CP hang where a rank with no supervised tokens produced a disconnected custom loss and therefore did not traverse the required backward collectives. We did not observe that failure in DSV4; AutoModel’s standard `MaskedCrossEntropy` was already shown to return a graph-connected zero when every label is ignored. The touch was added as a precaution rather than as a fix for a reproduced DSV4 issue.

The runtime scope was also narrow:

- Only DSV4 and GLM5.2 returned `_cp_full_logits_grad_touch`; no other model opted into it.
- DSV4 propagated the flag, but the consumer existed only in the non-PP training path.
- GLM5.2’s CP sharder reconstructs the batch from a fixed set of keys and drops the flag, so GLM5.2 never actually executed the touch.
- PP training never executed this block.

Validation after removal:

- Strengthened the all-masked `MaskedCrossEntropy` regression to verify a finite zero loss with `requires_grad=True`, followed by a non-`None`, all-zero logits gradient after backward.
- Ran the focused loss, DSV4, GLM5.2, and non-PP recipe tests: `41 passed`.
- Ruff and formatting checks passed.
- DSV4 CP1 packed THD completed `50/50` steps: [W&B](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/2n1dz0tv).
- DSV4 CP8 packed THD completed `100/100` steps: [W&B](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/2suf2cl1).
- GLM5.2 CP1 packed THD also completed successfully after the removal: [W&B](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/pssq9x60).

**Review by hemildesai — DISMISSED** [review]

File: nemo_automodel/_transformers/capabilities.py [hunk 3]
```diff
@@ -72,6 +72,12 @@ def _supports_seq_lens(model: "nn.Module") -> bool:
         return False
 
 
+def _supports_thd(model: "nn.Module") -> bool:
+    """True when the model capability declaration opts into native THD inputs."""
+    capabilities = getattr(model, "ModelCapabilities", None)
+    return bool(getattr(capabilities, "supports_thd", False))
```
Comment (akoumpa) [inline nemo_automodel/_transformers/capabilities.py]: @HuiyingLi is there a way to avoid introducing this function? In particular, I'm leaning towards relying more on types, so that we can avoid all the getattr.

File: nemo_automodel/_transformers/model_capabilities.py [hunk 4]
```diff
@@ -54,16 +54,18 @@ class ModelCapabilities:
         supports_cp: Context parallelism.
         supports_pp: Pipeline parallelism.
         supports_ep: Expert parallelism (MoE).
+        supports_thd: Native THD packed-sequence inputs.
```
Comment (akoumpa) [inline nemo_automodel/_transformers/model_capabilities.py]:
```suggestion
        supports_thd: THD packed-sequence inputs.
```

File: nemo_automodel/_transformers/model_capabilities.py [hunk 5]
```diff
@@ -54,16 +54,18 @@ class ModelCapabilities:
...
     supports_tp: bool = False
     supports_cp: bool = False
     supports_pp: bool = False
     supports_ep: bool = False
+    supports_thd: bool = False
 
 
 def _to_canonical(caps_obj) -> ModelCapabilities:
-    """Re-pack any object with the four ``supports_*`` fields into the canonical type.
+    """Re-pack a model capability declaration into the canonical type.
```
Comment (akoumpa) [inline nemo_automodel/_transformers/model_capabilities.py]: @HuiyingLi let's keep the original comment?

File: nemo_automodel/_transformers/model_capabilities.py [hunk 6]
```diff
@@ -76,6 +78,7 @@ def _to_canonical(caps_obj) -> ModelCapabilities:
         supports_cp=bool(caps_obj.supports_cp),
         supports_pp=bool(caps_obj.supports_pp),
         supports_ep=bool(caps_obj.supports_ep),
+        supports_thd=bool(getattr(caps_obj, "supports_thd", False)),
```
Comment (akoumpa) [inline nemo_automodel/_transformers/model_capabilities.py]:
```suggestion
        supports_thd=bool(caps_obj.supports_thd),
```
@HuiyingLi  can we do the above or is the getattr necessary?

File: nemo_automodel/components/distributed/cp_utils.py [hunk 7]
```diff
@@ -323,7 +323,7 @@ def _get_mesh_size(mesh):
     # the batch in its pre-embed step. Honor it instead of the default
     # load-balanced context_parallel path so the implementation stays with the model.
     cp_make_batch_fn = batch.pop("_cp_make_batch_fn", None)
```
Comment (akoumpa) [inline nemo_automodel/components/distributed/cp_utils.py:325]: Passing a function pointer via a dict makes debugging and type checking harder. Perhaps another PR should address this.

File: nemo_automodel/components/distributed/cp_utils.py [hunk 8]
```diff
@@ -323,7 +323,7 @@ def _get_mesh_size(mesh):
...
-    if _get_mesh_size(cp_mesh) > 1 and cp_make_batch_fn is not None:
+    if cp_make_batch_fn is not None:
```
Comment (akoumpa) [inline nemo_automodel/components/distributed/cp_utils.py:326]: @HuiyingLi do you know why the `_get_mesh_size(cp_mesh) > 1 and` is no-longer necessary?

Comment (HuiyingLi) [reply nemo_automodel/components/distributed/cp_utils.py:326]: thd is also enabled for cp=1 cases. Both models that takes this path: dsv4 and glm52 are tested for this enablement.

File: nemo_automodel/components/distributed/cp_utils.py (hunk 7 above)
Comment (HuiyingLi) [reply nemo_automodel/components/distributed/cp_utils.py:325]: Agree, this part is removed in https://github.com/NVIDIA-NeMo/Automodel/pull/2937

**Review by hemildesai — APPROVED** [review]
